Source/Core/Core/PowerPC/GDBThread.cpp

```diff
@@ -0,0 +1,1274 @@
+
+  t.tv_sec = 0;
+  t.tv_usec = 20;
+
+  if (select(m_sock + 1, fds, nullptr, nullptr, &t) < 0)
+    return -1;
+
+  if (FD_ISSET(m_sock, fds))
+    return 1;
+  return 0;
+}
+
+void GDBThread::GDBHandleEvents()
+{
+  if (!m_connected)
+    return;
+
+  while (0 < GDBDataAvailable())
+  {
+    GDBReadCommand();
+    GDBParseCommand();
+  }
+}
+
+void GDBThread::GDBReadCommand()
+{
+  m_cmd_len = 0;
+  memset(m_cmd_bfr, 0, sizeof m_cmd_bfr);
+
+  u8 c = GDBReadByte();
+  if (c != GDB_STUB_START)
+  {
+    ERROR_LOG(GDB_THREAD, "gdb: read invalid byte %02x\n", c);
+    return;
+  }
+
+  while ((c = GDBReadByte()) != GDB_STUB_END)
+  {
+    m_cmd_bfr[m_cmd_len++] = c;
+    if (m_cmd_len == sizeof m_cmd_bfr)
+      fail("gdb: cmd_bfr overflow\n");
+  }
+
+  u8 chk_read = HexToChar(GDBReadByte()) << 4;
+  chk_read |= HexToChar(GDBReadByte());
+
+  u8 chk_calc = GDBCalculateChecksum();
+
+  if (chk_calc != chk_read)
+  {
+    ERROR_LOG(GDB_THREAD,
+              "gdb: invalid checksum: calculated %02x and read %02x for $%s# (length: %d)\n",
+              chk_calc, chk_read, m_cmd_bfr, m_cmd_len);
+    m_cmd_len = 0;
+
+    GDBNak();
+  }
+
+  NOTICE_LOG(GDB_THREAD, "gdb: read command %c with a length of %d: %s\n", m_cmd_bfr[0], m_cmd_len,
+             m_cmd_bfr);
+}
+
+void GDBThread::GDBParseCommand()
+{
+  if (m_cmd_len == 0)
+    return;
+
+  switch (m_cmd_bfr[0])
+  {
+  case 'q':
+    GDBHandleQuery();
+    break;
+  case 'H':
+    GDBHandleSetThread();
+    break;
+  case '?':
+    GDBHandleSignal();
+    break;
+  case 'D':
+    GDBDetach();
+    break;
+  case 'k':
+    GDBKill();
+    break;
+  case 'g':
+    GDBReadRegisters();
+    break;
+  case 'G':
+    GDBWriteRegisters();
+    break;
+  case 'p':
+    GDBReadRegister();
+    break;
+  case 'P':
+    GDBWriteRegister();
+    break;
+  case 'm':
+    GDBReadMemory();
+    break;
+  case 'M':
+    GDBWriteMemory();
+    break;
+  case 'c':
+    GDBContinue();
+    break;
+  case 's':
+    GDBStep();
+    break;
+  case ' ':
+    GDBPause();
+    break;
+  case 'z':
+    GDBRemoveBreakpoint();
+    break;
+  case 'Z':
+    GDBAddBreakpoint();
+    break;
+  default:
+    GDBAck();
+    GDBReply("");
+    break;
+  }
+}
+
+u8 GDBThread::GDBReadByte()
+{
+  if (!m_connected)
+    return 0;
+
+  u8 c;
+
+  size_t res = recv(m_sock, reinterpret_cast<char*>(&c), 1, MSG_WAITALL);
+  if (res != 1)
+    failr("recv failed");
+
+  return c;
+}
+
+u8 GDBThread::GDBCalculateChecksum() const
+{
+  u32 len = m_cmd_len;
+  const u8* ptr = m_cmd_bfr;
+  u8 c = 0;
+
+  while (len-- > 0)
+    c += *ptr++;
+
+  return c;
+}
+
+void GDBThread::GDBReply(const char* reply)
+{
+  if (!m_connected)
+    return;
+
+  memset(m_cmd_bfr, 0, sizeof m_cmd_bfr);
+
+  m_cmd_len = static_cast<u32>(strlen(reply));
+  if (m_cmd_len + 4 > sizeof m_cmd_bfr)
+    fail("GDBReply: cmd_bfr overflow");
+
+  memcpy(m_cmd_bfr + 1, reply, m_cmd_len);
+
+  m_cmd_len++;
+  u8 chk = GDBCalculateChecksum();
+  m_cmd_len--;
```

Comment:
Just wondering why this place generates a checksum for the current command buffer size **plus one** but the other places don't.
It would probably be clearer if you straight up updated `m_cmd_len` after the `memcpy` (which does indeed skip the first byte and and start writing at the 2nd, assuming because it is some sort of command indicator) and continue with it for the remainder of the method.

But then again, you're also writing to `m_cmd_len + 3` (or +2 if you drop this decrease again) _without_ updating the length at all, so this is somewhat confusing (at least to me).

This probably _at least_ deserves a comment or two, for anyone that comes afterwards.

Reply:
Inherited from Dolphin's GDBStub implementation, which probably inherited it from the original C based gdbstub implementation. Will take a look.